File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:

    void expandFromCenter(string &s, int left, int right, int &start, int &maxLen)
    {
        int n = s.size();
        while (left >= 0 && right < n && s[left] == s[right]) {
            int currLen = right - left + 1;
            if (currLen > maxLen) {
                start = left;
                maxLen = currLen;
            }
            left--;
            right++;
        }
    }

    string longestPalindrome(string s) {
        int n = s.size();
        if (n == 0) return "";
        int start = 0;
        int maxLen = 1;
        for (int i = 0; i < n; i++) {
            expandFromCenter(s, i, i, start, maxLen);
            expandFromCenter(s, i, i + 1, start, maxLen);
        }

        return s.substr(start, maxLen);
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        start=0
        end=0
        for i in range(0,n):
            odd=self.expand(s,i,i)
            even=self.expand(s,i,i+1)
            length=max(odd,even)
            if(length>end-start):
                start=i-(length-1)//2
                end=i+length//2
        
        return s[start:end+1]

    def expand(self, s: str, left: int, right: int) -> int:
        while(left>=0 and right<len(s) and s[left]==s[right]):
            left-=1
            right+=1
        return right-left-1